File: 20231009_Monday/20231009.go
// Leetcode: 34. Find First and Last Position of Element in Sorted Array.
// https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 8 ms, faster than 67.04% of Go online submissions for Find First and
// Last Position of Element in Sorted Array.
// Memory Usage: 4 MB, less than 98.09% of Go online submissions for Find First
// and Last Position of Element in Sorted Array.
// 2023.10.09 Daily Challenge.
package main

var n []int

func searchRange(nums []int, target int) []int {
	// Based on:
	// https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/discuss/1054742/Python-O(logn)
	n = nums
	l, h := bs(target), bs(target+1)-1
	if l <= h {
		return []int{l, h}
	}
	return []int{-1, -1}
}

func bs(x int) int {
	l, h := 0, len(n)
	for l < h {
		m := l + (h-l)/2
		if n[m] < x {
			l = m + 1
		} else {
			h = m
		}
	}
	return l
}

File: 20231009_Monday/20231009.rb
# Leetcode: 34. Find First and Last Position of Element in Sorted Array.
# https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 70 ms, faster than 31.31% of Ruby online submissions for Find First
# and Last Position of Element in Sorted Array.
# Memory Usage: 211.5 MB, less than 82.83% of Ruby online submissions for Find
# First and Last Position of Element in Sorted Array.
# 2023.10.09 Daily Challenge.
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def search_range(nums, target)
  # Based on:
  # https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/discuss/1054742/Python-O(logn)
  @n = nums
  l = bs(target)
  h = bs(target+1) - 1
  return [l,h] if l <= h
  return [-1,-1]
end

def bs(x)
  lo, hi = 0, @n.size # ! [1]
  while lo < hi
    md = lo + (hi-lo)/2
    if @n[md] < x
      lo = md + 1
    else
      hi = md
    end
  end
  return lo
end
